### PR TITLE
[Draft for review] Push envs + tests

### PR DESCRIPTION
**Pull request by balsoft**
This is just for reviewing, we should merge the corresponding PRs instead of this

### CODE REVIEW

File: tests/modules/vault-secrets.nix [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  testScript = let
+    sshToServer =
+      "ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -i privkey.snakeoil server";
+
+      fakeFlake = {
+        nixosConfigurations.client = self.inputs.nixpkgs.lib.nixosSystem {
+          modules = [ nodes.client ];
+          inherit (pkgs) system;
+        };
+      };
```
Comment (mkaito) [inline tests/modules/vault-secrets.nix:46]: I have to admit it took me far too long to realize this is not nested under `sshToServer`, but just indented incorrectly.

File: tests/modules/vault-secrets.nix [hunk 2]
```diff
@@ -0,0 +1,78 @@
+{ nixosPath, self, ... }@args:
+
+(import "${nixosPath}/tests/make-test-python.nix" ({ pkgs, ... }: rec {
+  name = "vault-secrets";
+  nodes = {
+    server = { pkgs, ... }:
+      let
+        serverArgs =
+          "-dev -dev-root-token-id='root' -dev-listen-address='0.0.0.0:8200'";
```
Comment (mkaito) [inline tests/modules/vault-secrets.nix:9]: Wow, you can just hard code the token? I want the time back that I spent trying to actually transfer the generated token...

File: flake.nix [hunk 3]
```diff
@@ -1,12 +1,40 @@
...
 
-  };
+      overlay = final: prev: {
+        vault-push-approles =
+          final.callPackage ./scripts/vault-push-approles.nix { };
+        vault-push-approle-envs =
+          final.callPackage ./scripts/vault-push-approle-envs.nix { };
+      };
+      nixosModules.vault-secrets = import ./modules/vault-secrets.nix;
+    } // (flake-utils.lib.eachSystem [ "x86_64-linux" ] (system:
```
Comment (mkaito) [inline flake.nix:20]: Looping over a single item list :100:

Comment (balsoft) [reply flake.nix:20]: We will likely want to add `aarch64-linux` here soon, so might as well do this

Comment (mkaito) [reply flake.nix:20]: I guess we better get an aarch64 machine we can use as remote builder then.

File: scripts/vault-ensure-token.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+set -euo pipefail
+
+token_data="$(vault token lookup -format=json)"
+vault_token_never_expire="$(jq '.data.expire_time == null' <<< "$token_data")"
+vault_token_ttl="$(jq '.data.ttl' <<< "$token_data")"
+if [[ $vault_token_never_expire == false && $vault_token_ttl -le 0 ]]; then
+    echo 'Vault token expired or invalid. Please log into vault first.'
+    exit 1
+fi
```
Comment (mkaito) [inline scripts/vault-ensure-token.sh:11]: For what it's worth, note that this passage only ensures there is a token present and it has not expired. It makes no attempt at ensuring that the token is actually from the correct Vault remote, or that the token has the necessary access to perform the operations.

File: scripts/vault-push-approles.nix [hunk 5]
```diff
@@ -232,6 +221,7 @@
         lib.concatMapStringsSep "\n" writeApprole allApproleParams;
     in writeShellScriptBin "vault-push-approles" ''
       set -euo pipefail
+      export PATH=$PATH''${PATH:+':'}'${jq}/bin:${vault}/bin:${coreutils}/bin:${bash}/bin'
```
Comment (mkaito) [inline scripts/vault-push-approles.nix:224]: Wasn't there a thing like `mkBinPath`?

File: tests/modules/vault-secrets.nix [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        let
+          serverArgs =
+            "-dev -dev-root-token-id='root' -dev-listen-address='0.0.0.0:${toString vault-port}'";
+        in {
+          # An unsealed dummy vault
+          networking.firewall.allowedTCPPorts = [ vault-port ];
+          systemd.services.dummy-vault = {
+            wantedBy = [ "multi-user.target" ];
+            path = with pkgs; [ getent vault ];
+            script = "vault server ${serverArgs}";
```
Comment (mkaito) [inline tests/modules/vault-secrets.nix:21]: Why wrap it in an unnecessary shell process when you could just `ExecStart` directly?

File: scripts/vault-push-approle-envs.nix [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+      # Find all configurations that have vault-secrets defined
+      configsWithSecrets = lib.filterAttrs (_: cfg:
+        cfg.config ? vault-secrets && cfg.config.vault-secrets.secrets != { })
+        nixosConfigurations;
+
+      # Get all approles for all NixOS configurations in the given flake
+      approleParamsForAllMachines =
+        builtins.mapAttrs approleParamsForMachine configsWithSecrets;
+
+      # All approles for all NixOS configurations plus the extra approles
```
Comment (mkaito) [inline scripts/vault-push-approle-envs.nix:73]: What extra approles

File: scripts/vault-push-approle-envs.nix [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+              traversed = [ ];
+              result = true; # In an empty list, all elements are unique
+            };
+        in (allUnique' lst).result;
+
+      # A script to write all approles
+      pushAllApproleEnvs =
+        assert allUnique (map (x: x.approleName) allApproleParams);
+        lib.concatMapStringsSep "\n" pushApproleEnv allApproleParams;
+    in writeShellScriptBin "vault-push-approle-envs" ''
```
Comment (mkaito) [inline scripts/vault-push-approle-envs.nix:98]: Is it missing a `let` here?

File: flake.nix (hunk 3 above)
Comment (balsoft) [reply flake.nix:20]: Albali could do this with `binfmt.emulatedSystems = ["x86_64-linux"];`

File: scripts/vault-push-approles.nix (hunk 5 above)
Comment (balsoft) [reply scripts/vault-push-approles.nix:224]: Yeah, right

File: scripts/vault-push-approle-envs.nix (hunk 7 above)
Comment (balsoft) [reply scripts/vault-push-approle-envs.nix:73]: Copy-paste

File: scripts/vault-ensure-token.sh (hunk 4 above)
Comment (balsoft) [reply scripts/vault-ensure-token.sh:11]: I think that's just fine -- this is just a "fail-early" thing for when you forgot to authenticate

File: tests/modules/vault-secrets.nix (hunk 6 above)
Comment (balsoft) [reply tests/modules/vault-secrets.nix:21]: Because `path` only applies to `script`

**Review by mkaito — APPROVED** [review]

**Comment by mkaito** [comment]
So I think we're done here?

**Pull request closed without merging**